src/tools/miri/src/intptrcast.rs

```diff
@@ -158,7 +148,12 @@ impl<'mir, 'tcx> GlobalStateInner {
             ProvenanceMode::Permissive => {}
         }
 
-        // This is how wildcard pointers are born.
+        // We do *not* look up the `AllocId` here! This is a `ptr as usize` cast, and it is
+        // completely legal to do a cast and then `wrapping_offset` to another allocation and only
+        // *then* do a memory access. So the allocation that the pointer happens to point to on a
+        // cast is fairly irrelevant. Instead we generate this as a "wildcard" point, such that
+        // *every time the pointer is used*, we do an `AllocId` lookup to find the (exposed)
+        // allocation it might be referencing.
```

Comment:
I do not have any objection to the rest of this PR, but this is by far the most helpful. I'm not sure we really explain that this is how wildcard pointers behave anywhere else.

Reply:
Ah, good point! Yes indeed we entirely forgot about that.

I added a doc comment to `Pointer::Wildcard` that explains the general principle even more.